daemon/containerd/mount.go

```diff
@@ -30,8 +31,21 @@ func (i *ImageService) Mount(ctx context.Context, container *container.Container
 
 // Unmount unmounts the container base filesystem
 func (i *ImageService) Unmount(ctx context.Context, container *container.Container) error {
+	if container.BaseFS == "" {
+		target, err := i.refCountMounter.Mounted(container.ID)
+		if err != nil {
+			log.G(ctx).WithField("containerID", container.ID).Warn("failed to determine if container is already mounted")
+		}
+		if target != "" {
+			container.BaseFS = target
+		}
```

Comment:
The `if target != "" {` check is redundant here (we only reach this code if `container.BaseFS == "" `, so setting it to an empty string again, would be idempotent.

I'm more curious though; is `target` being empty a _valid_ situation, or should it produce an error?

Reply:
Oh, actually, it looks like we check for that below, so should we return immediately here? (Or should we even have a utility doing all this, that returns an error when empty? (Perhaps that may be too much though)

Reply:
`target == ""` is different from `err != nil` because it means that the target is not mounted, but there was no error determining that.

Reply:
Was thinking along these lines; basically we only need to check for empty-string in the `if container.BaseFS == "" {` branch;

```go
// Unmount unmounts the container base filesystem
func (i *ImageService) Unmount(ctx context.Context, container *container.Container) error {
	if container.BaseFS == "" {
		target, err := i.refCountMounter.Mounted(container.ID)
		if err != nil {
			log.G(ctx).WithField("containerID", container.ID).Warn("failed to determine if container is already mounted")
		}
		if target == "" {
			return errors.New("BaseFS is empty")
		}
		container.BaseFS = target
	}
	root := container.BaseFS
	if err := i.refCountMounter.Unmount(root); err != nil {
		log.G(ctx).WithField("container", container.ID).WithError(err).Error("error unmounting container")
		return fmt.Errorf("failed to unmount %s: %w", root, err)
	}

	return nil
}
```

Also wondering if we need the intermediate variables (`target` and `root`), but perhaps we don't want to update the `container` (as it's passed by reference); not sure?

```go
// Unmount unmounts the container base filesystem
func (i *ImageService) Unmount(ctx context.Context, container *container.Container) error {
	if container.BaseFS == "" {
		var err error
		container.BaseFS, err = i.refCountMounter.Mounted(container.ID)
		if err != nil {
			log.G(ctx).WithField("containerID", container.ID).Warn("failed to determine if container is already mounted")
		}
		if container.BaseFS == "" {
			return errors.New("BaseFS is empty")
		}
	}
	if err := i.refCountMounter.Unmount(container.BaseFS); err != nil {
		log.G(ctx).WithField("container", container.ID).WithError(err).Error("error unmounting container")
		return fmt.Errorf("failed to unmount %s: %w", container.BaseFS, err)
	}

	return nil
}
```

Reply:
Ah, thanks!

Went for the first option - I think having an intermediate variable is a bit more readable and doesn't require explicitly declaring an `err` variable.

Reply:
Yup, no problem with that. Always going back and forth between those myself, and depends a bit on a "per case". Thanks!